src/Features/Core/Diagnostics/EngineV2/DiagnosticIncrementalAnalyzer.cs

```diff
@@ -137,87 +230,446 @@ public override Task<ImmutableArray<DiagnosticData>> GetSpecificCachedDiagnostic
+        private class CompilationResult
         {
-            var groups = diagnostics.GroupBy(d => d.DocumentId);
+            public VersionStamp ProjectVersion { get; private set; }
+
+            private readonly ConcurrentDictionary<DocumentId, ImmutableArray<Diagnostic>> _diagnosticsPerDocuments = new ConcurrentDictionary<DocumentId, ImmutableArray<Diagnostic>>();
+            private readonly ConcurrentDictionary<DocumentId, Task> _analyzedDocuments = new ConcurrentDictionary<DocumentId, Task>();
+            private readonly object _updateDiagnosticsLock = new object();
+            private readonly DocumentId _noDocument;
+            private CompilationWithAnalyzers _compilationWithAnalyzers;
+            private Project _project;
 
-            var solution = project.Solution;
-            var workspace = solution.Workspace;
+            private Task _completionTask;
+            private bool _completionCanceled;
 
-            foreach (var kv in groups)
+            public CompilationResult(Project project, VersionStamp projectVersion, CompilationWithAnalyzers compilation)
             {
-                if (kv.Key == null)
+                _compilationWithAnalyzers = compilation;
+                _project = project;
+                this.ProjectVersion = projectVersion;
+
+                if (project != null)
                 {
-                    Owner.RaiseDiagnosticsUpdated(
-                        this, new DiagnosticsUpdatedArgs(
-                            ValueTuple.Create(this, project.Id), workspace, solution, project.Id, null, kv.ToImmutableArrayOrEmpty()));
-                    continue;
+                    _noDocument = new DocumentId(project.Id, Guid.Empty, "No document");
+                }
+            }
+
+            public bool CompletionStarted
+            {
+                get { return _completionTask != null; }
+            }
+
+            public bool DocumentAnalyzed(DocumentId documentId)
+            {
+                return _analyzedDocuments.ContainsKey(documentId);
+            }
+
+            public async Task<ImmutableArray<Diagnostic>> AnalyzeProjectAsync(CancellationToken cancellationToken)
+            {
+                ImmutableArray<Diagnostic> diagnostics = ImmutableArray<Diagnostic>.Empty;
+                if (_compilationWithAnalyzers != null)
+                {
+                    lock (_updateDiagnosticsLock)
+                    {
+                        if (!this.CompletionStarted)
+                        {
+                            _completionTask = Task.Run(async () =>
+                            {
+                                // If any documents have been separately analyzed, complete analysis of the project
+                                // by analyzing each document individually. Otherwise, analyze the project as a whole.
+                                if (_analyzedDocuments.Count > 0)
+                                {
```

Comment:
If you take a peek at CSharpCompilation.CompleteTree or VisualBasicCompilation.CompleteTree, the trigger that causes the compiler to decide that the compilation is complete (at least in terms of compilation events) is analysis of the last tree.